library/src/main/java/com/datatorrent/lib/state/managed/Bucket.java

```diff
@@ -0,0 +1,534 @@
+                readCache.put(key, bucketedValue);
+                return valSlice;
+              }
+            }
+          }
+
+        } catch (IOException e) {
+          throw new RuntimeException("get time-buckets " + bucketId, e);
+        }
+      }
+      return null;
+    }
+
+    @Override
+    public Slice get(Slice key, long timeBucket, ReadSource readSource)
+    {
+      switch (readSource) {
+        case MEMORY:
+          return getFromMemory(key);
+        case READERS:
+          return getFromReaders(key, timeBucket);
+        case ALL:
+        default:
+          Slice value = getFromMemory(key);
+          if (value != null) {
+            return value;
+          }
+          return getFromReaders(key, timeBucket);
+      }
+    }
+
+    /**
+     * Returns the value for the key from a time-bucket reader
+     * @param key        key
+     * @param timeBucket time bucket
+     * @return value if key is found in the time bucket; false otherwise
+     */
+    private Slice getKeyFromTimeBucketReader(Slice key, long timeBucket)
+    {
+      FileAccess.FileReader fileReader = readers.get(timeBucket);
+      if (fileReader != null) {
+        return readKey(fileReader, key, timeBucket);
+      }
+      //file reader is not loaded and is null
+      try {
+        if (loadFileReader(timeBucket)) {
+          return readKey(readers.get(timeBucket), key, timeBucket);
+        }
+        return null;
+      } catch (IOException e) {
+        throw new RuntimeException("while loading " + bucketId + ", " + timeBucket, e);
+      }
+    }
+
+    private Slice readKey(FileAccess.FileReader fileReader, Slice key, long timeBucket)
+    {
+      Slice valSlice = new Slice(null, 0, 0);
+      try {
+        if (fileReader.seek(key)) {
+          fileReader.next(dummyGetKey, valSlice);
+          return valSlice;
+        } else {
+          return null;
+        }
+      } catch (IOException e) {
+        throw new RuntimeException("reading " + bucketId + ", " + timeBucket, e);
+      }
+    }
+
+    private boolean loadFileReader(long timeBucketId) throws IOException
+    {
+      BucketsFileSystem.ImmutableTimeBucketMeta tbm = managedStateContext.getBucketsFileSystem()
+          .getTimeBucketMeta(bucketId, timeBucketId);
+
+      if (tbm != null) {
+        FileAccess.FileReader reader = managedStateContext.getBucketsFileSystem().getReader(bucketId,
+            BucketsFileSystem.getFileName(timeBucketId));
+        readers.put(timeBucketId, reader);
+        sizeInBytes.getAndAdd(tbm.getSizeInBytes());
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public void put(Slice key, long timeBucket, Slice value)
+    {
+      BucketedValue bucketedValue = flash.get(key);
+      if (bucketedValue == null) {
+        bucketedValue = new BucketedValue();
+        flash.put(key, bucketedValue);
```

Comment:
A put operation can modify flash in one thread while an asynchronous get operation can do a get from flash in another thread. Could this cause a problem?